runtime/context.go

```diff
@@ -271,3 +296,97 @@ func (c *contextLogger) Warn(ctx context.Context, msg string, userFields ...zap.
 func (c *contextLogger) Check(lvl zapcore.Level, msg string) *zapcore.CheckedEntry {
 	return c.log.Check(lvl, msg)
 }
+
+// ContextMetrics emits metrics.
+type ContextMetrics interface {
+	GetOrAddInboundHTTPMetrics(ctx context.Context) *InboundHTTPMetrics
+	GetOrAddEndpointMetrics(ctx context.Context) *EndpointMetrics
+	GetOrAddInboundTChannelMetrics(ctx context.Context) *InboundTChannelMetrics
+	getEndpointMetrics(key string) *EndpointMetrics
+	getInboundHTTPMetrics(key string) *InboundHTTPMetrics
+	getInboundTChannelMetrics(key string) *InboundTChannelMetrics
+}
+
+type contextMetrics struct {
+	sync.RWMutex
+	scope                  tally.Scope
+	endpointMetrics        map[string]*EndpointMetrics
+	inboundHTTPMetrics     map[string]*InboundHTTPMetrics
+	inboundTChannelMetrics map[string]*InboundTChannelMetrics
+}
+
+// NewContextMetrics emits metrics.
+func NewContextMetrics(scope tally.Scope) ContextMetrics {
+	return &contextMetrics{
+		scope:                  scope,
+		endpointMetrics:        make(map[string]*EndpointMetrics),
+		inboundHTTPMetrics:     make(map[string]*InboundHTTPMetrics),
+		inboundTChannelMetrics: make(map[string]*InboundTChannelMetrics),
+	}
+}
+
+// GetOrAddInboundHTTPMetrics add tags to scope and create inbound http metrics
+func (c *contextMetrics) GetOrAddInboundHTTPMetrics(ctx context.Context) *InboundHTTPMetrics {
+	tags := GetRequestScopeFieldsFromCtx(ctx)
+	key := tally.KeyForStringMap(tags)
+	c.RLock()
+	if val, ok := c.inboundHTTPMetrics[key]; ok {
+		c.RUnlock()
+		return val
+	}
+
+	c.RUnlock()
+	scope := c.scope.Tagged(tags)
+	c.Lock()
+	c.inboundHTTPMetrics[key] = NewInboundHTTPMetrics(scope)
+	c.Unlock()
+	return c.inboundHTTPMetrics[key]
+}
+
+// GetOrAddEndpointMetrics add tags to scope and create endpoint metrics
+func (c *contextMetrics) GetOrAddEndpointMetrics(ctx context.Context) *EndpointMetrics {
+	tags := GetEndpointScopeFieldsFromCtx(ctx)
+	key := tally.KeyForStringMap(tags)
+	c.RLock()
+	if val, ok := c.endpointMetrics[key]; ok {
+		c.RUnlock()
+		return val
+	}
+
+	c.RUnlock()
+	scope := c.scope.Tagged(tags)
+	c.Lock()
```

Comment:
Do we know what kind of tags there will be and how dynamic they are? If the tags are very dynamic, then we will go through this write lock path a lot, and that could potentially be problematic performance wise. But if the tags are rather static or don't vary a lot across requests, we can probably live with it. Let's figure out which case it is. We should definitely run some benchmarks to see the impact of these locks.

Reply:
Add a bench mark testing file and my local results seems not very different from without tags one:

```
xuan:zanzibar xuan.he$ wrk -t25 -c400 -d300s -s ./benchmarks/contacts_2500KB.lua http://localhost:8093/contacts/foo/contacts
Running 5m test @ http://localhost:8093/contacts/foo/contacts
  25 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    14.36ms   13.07ms 394.41ms   78.59%
    Req/Sec   814.47    411.89     2.27k    57.09%
  2917600 requests in 5.00m, 320.02MB read
  Socket errors: connect 174, read 337, write 0, timeout 1
  Non-2xx or 3xx responses: 972
Requests/sec:   9722.23
Transfer/sec:      1.07MB

```

I feel it is hard to fully simulate the scenario in production since we don't know the fashion of when to create a new key in the metrics map which needs write lock